### ISSUE TITLE
Malformed UTF8

### ISSUE THREAD

**Issue by ant1j** [curtisalexander/readstat-rs#78]
Hello,

While converting some huge sas file, I encountered some error linked to UTF8 decoding error.
```thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Utf8Error { valid_up_to: 3, error_len: None }', crates\readstat\src\common.rs:40:36```

The function is:
```
pub unsafe fn ptr_to_string(x: *const i8) -> String {
    if x.is_null() {
        String::new()
    } else {
        CStr::from_ptr(x).to_str().unwrap().to_owned() // <-- here
    }
}
```

The issue came from the fact that some values were truncated in the middle of an UTF8 sequence (e.g. `println!("{:#?}",CStr::from_ptr(x));` showed `"CAF\xc3"` for example.

Since I am not at all coding in rust, I came up with this ugly work around, using the `CStr::from_ptr(x).to_string_lossy()` method:
```
pub unsafe fn ptr_to_string(x: *const i8) -> String {
    if x.is_null() {
        String::new()
    } else {
        // CStr::from_ptr(x).to_str().unwrap().to_owned()
        let val = CStr::from_ptr(x);
        let strval = val.to_str();
        if strval.is_err() {
            println!(
                "Error while converting to UTF8: {:#?}. Will use lossy conversion instead.",
                val
            );
            val.to_string_lossy().into_owned()
        } else {
            strval.unwrap().to_owned()
        }
    }
}
```

Could it make sense to have additional flag (`--lossy-utf8` ?) to avoid these errors?

**Comment by curtisalexander** [comment]
@ant1j - First, thank you for the excellent report and diagnosis.  Admittedly I have too many `unwrap` statements in the Rust code and not all errors are handled gracefully.  Definitely something that can be improved.

Yes, I would consider adding a flag for lossy UTF8.  Another possibility would be to mimic the example you outlined.  Instead of performing the `unwrap` that is in place now, check for an error and if there is an issue 1) perform a lossy conversion and 2) write a warning to standard error (rather than standard out).

Would you be able to share a SAS file (even just a few rows/columns) that contains malformed UTF8?  I definitely don’t want anything that is proprietary or contains sensitive information.  I would use the example for testing the change and would include in the suite of integration tests I have against SAS files.

**Comment by curtisalexander** [comment]
@ant1j - Can you try branch #79 to see if that fixes your issue?

**Comment by ant1j** [comment]
I am not able to share any `sas7bdat` file because I dont have access to a SAS instance myself unfrotunately.

My best guess is that during initial export of the file (before entering SAS) or at SAS import, the field was truncated to a length of 4 (i.e 4 bytes I guess) and thus been cut in the middle of a multi-byte UTF8 char sequence - the raw display of the faulty variables show that they all ended with the byte code that starts accentuated chars (`0x62` or `0x63` if I remember correctly).

Release `0.12.2` worked perfeclty under WSL (I usually work on Windows, but can also use WSL) for a 8.5m rows file that was faulty before.

**Comment by curtisalexander** [comment]
@ant1j - No problem about the SAS file.  I am going to merge the update into `main`.

One thing I did not implement was writing to standard error if a lossy conversion is performed.  I'm not sure the added logic to check for lossy conversion is particularly helpful.  Within the result set one should be able to see that a lossy conversion occurred via the presence of the [replacement character](https://doc.rust-lang.org/std/char/constant.REPLACEMENT_CHARACTER.html).

**Comment by ant1j** [comment]
Sure, agreed.
Thanks for that.

You are not providing windows build anymore?

**Comment by curtisalexander** [comment]
No, I did not mean to leave out a Windows build.  Let me retrigger GitHub Actions.

**Comment by curtisalexander** [comment]
@ant1j - Fixed - Windows build now available.
